pipeline/mutate/mutator_noop.go

```diff
@@ -23,7 +23,22 @@ func (a *MutatorNoop) GetID() string {
 }
 
 func (a *MutatorNoop) Mutate(r *http.Request, session *authn.AuthenticationSession, config json.RawMessage, _ pipeline.Rule) error {
+	currentSessionHeaders := session.Header.Clone()
 	session.Header = r.Header
+	if session.Header == nil {
+		session.Header = make(map[string][]string)
+	}
+
+	for k, v := range currentSessionHeaders {
+		var val string
+		if len(v) == 0 {
+			val = ""
+		} else {
+			val = v[0]
+		}
+		session.SetHeader(k, val)
+	}
+
```

Comment:
Thank you for the fix - the bug you encounter for sure is annoying.

I am trying to figure out what's going on here. So `session.Header` was (I think) originally used by mutators to set a custom header for the upstream. Similarly, the decision endpoint also wants to set these headers (e.g. a JSON Web Token) so that the proxy can forward them to the upstream.

I think that you'll actually have the same problem with other mutators as well, because they are not designed to handle headers from the authorization piece.

It probably makes sense to manage merging the two headers from the incoming request and the session outside of the handlers?

And it definitely needs a failing test case first to ensure this fix works and doesn't regress.

Reply:
Sorry for not getting back to your review quickly. I believe I looked at the code for the other mutators and they didn't have that problem. I'll have another look and see if I can find an elegant way to handle it outside of the handlers if that makes sense. Initially I made this PR as small as possible so it would be easy to review. I've deployed this fix already so I know it works, but a test would be good to add to ensure it doesn't regress in the future.

Reply:
@zepatrik Maybe you can way in here as well.